Andrea del Pilar Zamora Cortes 07/09/2023

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

1. Criterio de Optimización:
Minimizar el costo total de la comida para gatos.

2. Variables de Control:
Catidad de cada ingrediente
- $x_1$: Gramos de pollo.
- $x_2$: Gramos de carne.
- $x_3$: Gramos de cordero.
- $x_4$: Gramos de arroz.
- $x_5$: Gramos de trigo.
- $x_6$: Gramos de gel.

3. Función de Costo:
Suma de los costos de los ingredientes
$$0.013x_1+0.008x_2+0.010x_3+0.002x_4+0.005x_5+0.001x_6$$


4. Restricciones:

Nutricionales:

Proteína: $0.1x_1+0.2x_2+0.15x_3+0x_4+0.04x_5+0x_6\ge 8$

Grasa: $0.08x_1+0.1x_2+0.11x_3+0.01x_4+0.01x_5+0x_6\ge 6 $

Fibra: $0.001x_1+0.005x_2+0.005x_3+0.1x_4+0.15x_5+0x_6\le 2$

Sal: $0.002x_1+0.005x_2+0.007x_3+0.002x_4+0.008x_5+0x_6\le 0.4$

Cantidad: $x_1+x_2+x_3+x_4+x_5+x_6 = 100$

In [2]:
import scipy.optimize as opt


In [5]:
import numpy as np

c = np.array([0.013,
              0.008,
              0.01,
              0.002,
              0.005,
              0.001])
Aub = np.array([
    [-0.1, -0.2, -0.15, 0, -0.04, 0],
    [-0.08, -0.1, -0.11, -0.01, -0.01, 0],
    [0.001, 0.005, 0.005, 0.10, 0.15, 0],
    [0.002, 0.005, 0.007, 0.002, 0.008, 0]
])
bub = np.array([-8,
                -6,
                 2,
                0.4])
Aeq = np.array([
    [1, 1, 1, 1, 1, 1]
])
beq = np.array([100])

sol=opt.linprog(c, Aub, bub, Aeq, beq)
sol

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 0.52
              x: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00  0.000e+00
                   4.000e+01]
            nit: 2
          lower:  residual: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00
                              0.000e+00  4.000e+01]
                 marginals: [ 6.400e-03  0.000e+00  1.300e-03  3.000e-04
                              3.300e-03  0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: [ 0.000e+00]
                 marginals: [ 1.000e-03]
        ineqlin:  residual: [ 4.000e+00  0.000e+00  1.700e+00  1.000e-01]
                 marginals: [-0.000e+00 -7.000e-02 -0.000e+00 -0.000e+00]
 mip

In [6]:
sol.x

array([ 0., 60.,  0.,  0.,  0., 40.])

In [7]:
sol.fun

0.52

En 60 gramos de carne y 40 gramos de gel se tiene 12 gramos de proteina, 6 gramos de grasa y 0.3 gramos fibra y sal, con un costo de 0.52.

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

1. Criterio de Optimización:
Minimizar los costos totales asociados a la distribución de energía eléctrica

2. Variables de Control:
Cantidades de energía eléctrica que se distribuyen desde cada planta a cada ciudad
- $x_1$: Kw distribuidos de la Planta 1 a Guadalajara
- $x_2$: Kw distribuidos de la Planta 1 a León
- $x_3$: Kw distribuidos de la Planta 1 a Morelia
- $x_4$: Kw distribuidos de la Planta 2 a Guadalajara
- $x_5$: Kw distribuidos de la Planta 2 a León
- $x_6$: Kw distribuidos de la Planta 2 a Morelia
- $x_7$: Kw distribuidos de la Planta 3 a Guadalajara
- $x_8$: Kw distribuidos de la Planta 3 a León
- $x_9$: Kw distribuidos de la Planta 3 a Morelia

3. Función de Costo:
Suma de los costos de los ingredientes
$$ 5𝑥_1 + 2𝑥_2 + 7𝑥_3 + 3𝑥_4 + 6𝑥_5 + 6𝑥_6 + 6𝑥_7 + 1𝑥_8 + 2𝑥_9$$-


4. Restricciones:

- Capacidad de plantas:

Planta 1: $𝑥_1 + 𝑥_2 + 𝑥_3\le 80$

Planta 2: $𝑥_4 + 𝑥_5 + 𝑥_6\le 40$ 

Planta 3: $𝑥_7 + 𝑥_8 + 𝑥_9\le 60$

- Demanda de las ciudades:

Guadalajara: $𝑥_1 + 𝑥_4 + 𝑥_7\ge 70$ 

León: $𝑥_2 + 𝑥_5 + 𝑥_8\ge 40$

Morelia: $𝑥_3 + 𝑥_6 + 𝑥_9\ge 70$ 

In [29]:
import numpy as np

c = np.array([5,
              2,
              7,
              3,
              6,
              6,
              6,
              1,
              2])
Aub = np.array([
    [-1, -1, -1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, -1, -1, -1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, -1, -1, -1],
    [1, 0, 0, 1, 0, 0, 1, 0, 0],
    [0, 1, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 1, 0, 0, 1, 0, 0, 1]
])
bub = np.array([-80,
                -40,
                -60,
                 70,
                 40,
                 70])

sol=opt.linprog(c, Aub, bub)
sol

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 540.0
              x: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01  0.000e+00
                   0.000e+00  0.000e+00  0.000e+00  6.000e+01]
            nit: 6
          lower:  residual: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01
                              0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+01]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+00  1.000e+00  6.000e+00  4.000e+00
                              0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0.0

In [30]:
sol.x

array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])

In [31]:
sol.fun

540.0

In [32]:
C=30*5+40*2+10*7+40*3+60*2
C

540

In [33]:
Guadalajara=30*5+40*3
Guadalajara

270

In [35]:
Leon= 40*2
Leon

80

In [37]:
Morelia= 10*7+60*2
Morelia

190

Por lo tanto la planta 1 usa el 80 millones de KW, planta 2 da 40 millones de KW y las planta 3 tiene 70 millones de KW, con un costo total de 540